server/src/DBTablesLastInfo.cc

```diff
@@ -0,0 +1,484 @@
+{
+	return TABLE_NAME_LAST_INFO;
+}
+
+void DBTablesLastInfo::reset(void)
+{
+	getSetupInfo().initialized = false;
+}
+
+const DBTables::SetupInfo &DBTablesLastInfo::getConstSetupInfo(void)
+{
+	return getSetupInfo();
+}
+
+DBTablesLastInfo::DBTablesLastInfo(DBAgent &dbAgent)
+: DBTables(dbAgent, getSetupInfo()),
+  m_impl(new Impl())
+{
+}
+
+DBTablesLastInfo::~DBTablesLastInfo()
+{
+}
+
+LastInfoIdType DBTablesLastInfo::addLastInfo(LastInfoDef &lastInfoDef,
+                                             const OperationPrivilege &privilege)
+{
+	HatoholError err = checkPrivilegeForAdd(privilege, lastInfoDef);
+	if (err != HTERR_OK)
+		return INVALID_LAST_INFO_ID;
+
+	LastInfoIdType lastInfoId;
+	DBAgent::InsertArg arg(tableProfileLastInfo);
+	arg.add(AUTO_INCREMENT_VALUE);
+	arg.add(lastInfoDef.dataType);
+	arg.add(lastInfoDef.value);
+	arg.add(lastInfoDef.serverId);
+
+	getDBAgent().runTransaction(arg, &lastInfoId);
+	return lastInfoId;
+}
+
+HatoholError DBTablesLastInfo::updateLastInfo(LastInfoDef &lastInfoDef,
+                                            const OperationPrivilege &privilege)
+{
+	HatoholError err = checkPrivilegeForUpdate(privilege, lastInfoDef);
+	if (err != HTERR_OK)
+		return err;
+
+	DBAgent::UpdateArg arg(tableProfileLastInfo);
+
+	const char *lastInfoIdColumnName =
+	  COLUMN_DEF_LAST_INFO[IDX_LAST_INFO_ID].columnName;
+	arg.condition = StringUtils::sprintf("%s=%" FMT_LAST_INFO_ID,
+	                                     lastInfoIdColumnName, lastInfoDef.id);
+	arg.add(IDX_LAST_INFO_DATA_TYPE, lastInfoDef.dataType);
+	arg.add(IDX_LAST_INFO_VALUE, lastInfoDef.value);
+	arg.add(IDX_LAST_INFO_SERVER_ID, lastInfoDef.serverId);
+
+	getDBAgent().runTransaction(arg);
+	return HTERR_OK;
+}
+
+HatoholError DBTablesLastInfo::getLastInfoList(LastInfoDefList &lastInfoDefList,
+					       const LastInfoQueryOption &option)
+{
+
+	DBAgent::SelectExArg arg(tableProfileLastInfo);
+	arg.add(IDX_LAST_INFO_ID);
+	arg.add(IDX_LAST_INFO_DATA_TYPE);
+	arg.add(IDX_LAST_INFO_VALUE);
+	arg.add(IDX_LAST_INFO_SERVER_ID);
+
+	// condition
+	arg.condition = option.getCondition();
+
+	getDBAgent().runTransaction(arg);
+
+	const ItemGroupList &grpList = arg.dataTable->getItemGroupList();
+	ItemGroupListConstIterator itemGrpItr = grpList.begin();
+	for (; itemGrpItr != grpList.end(); ++itemGrpItr) {
+		ItemGroupStream itemGroupStream(*itemGrpItr);
+		LastInfoDef lastInfoDef;
+
+		itemGroupStream >> lastInfoDef.id;
+		itemGroupStream >> lastInfoDef.dataType;
+		itemGroupStream >> lastInfoDef.value;
+		itemGroupStream >> lastInfoDef.serverId;
+
+		lastInfoDefList.push_back(lastInfoDef);
+	}
+
+	return HTERR_OK;
+}
+
+static string makeIdListCondition(const LastInfoIdList &idList)
+{
+	string condition;
+	const ColumnDef &colId = COLUMN_DEF_LAST_INFO[IDX_LAST_INFO_ID];
+	SeparatorInjector commaInjector(",");
+	condition = StringUtils::sprintf("%s in (", colId.columnName);
+	LastInfoIdListConstIterator it = idList.begin();
+	while (true) {
```

Comment:
I like later one style.

But, we can use more sophisticated range-based for style:

```cpp
for (auto id : idList) {
	commaInjector(condition);
	condition += StringUtils::sprintf("%ld", id);
}
```

How about this one?